velox/functions/prestosql/json/JsonStringUtil.cpp

```diff
@@ -0,0 +1,163 @@
+    const unsigned char* end) {
+  if (string == end) {
+    return 0;
+  }
+  if ((*string & 0x80u) == 0) {
+    return 1;
+  }
+  if ((*string & 0xE0u) == 0xC0u && (string + 1) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u) {
+    return 2;
+  }
+  if ((*string & 0xF0u) == 0xE0u && (string + 2) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u && (*(string + 2) & 0xC0u) == 0x80u) {
+    return 3;
+  }
+  if ((*string & 0xF8u) == 0xF0u && (string + 3) < end &&
+      (*(string + 1) & 0xC0u) == 0x80u && (*(string + 2) & 0xC0u) == 0x80u &&
+      (*(string + 3) & 0xC0u) == 0x80u) {
+    return 4;
+  }
+  return -1;
+}
+
+FOLLY_ALWAYS_INLINE char hexDigit(uint8_t c) {
+  return c < 10 ? c + '0' : c - 10 + 'A';
+}
+
+FOLLY_ALWAYS_INLINE void writeHex(char16_t value, char*& out) {
+  *out++ = '\\';
+  *out++ = 'u';
+  *out++ = hexDigit((value >> 12) & 0x0F);
+  *out++ = hexDigit((value >> 8) & 0x0F);
+  *out++ = hexDigit((value >> 4) & 0x0F);
+  *out++ = hexDigit(value & 0x0F);
+}
+
+std::array<int8_t, 128> getAsciiEscapes() {
+  std::array<int8_t, 128> escapes;
+  std::fill(escapes.data(), escapes.data() + 32, -1);
+  escapes['"'] = '"';
+  escapes['\\'] = '\\';
+  escapes['\b'] = 'b';
+  escapes['\t'] = 't';
+  escapes['\n'] = 'n';
+  escapes['\f'] = 'f';
+  escapes['\r'] = 'r';
+  return escapes;
+}
+static const std::array<int8_t, 128> asciiEscapes = getAsciiEscapes();
+
+FOLLY_ALWAYS_INLINE void encodeAscii(uint8_t value, char*& out) {
+  int8_t escapeCode = asciiEscapes[value];
+  if (escapeCode == 0) {
+    *out++ = char(value);
+  } else if (escapeCode > 0) {
+    *out++ = '\\';
+    *out++ = char(escapeCode);
+  } else {
+    writeHex(value, out);
+  }
+}
+
+/// Flush the chars in `buffer` till `end` to the `output` string. Reset `end`
+/// to the beginning of `buffer`.
+FOLLY_ALWAYS_INLINE void flush(char* buffer, char*& end, std::string& output) {
+  output.append(buffer, end - buffer);
+  end = buffer;
+}
+
+} // namespace
+
+void encodeUtf16Hex(char32_t codePoint, char*& out) {
+  VELOX_DCHECK(codePoint <= 0x10FFFFu);
+  if (codePoint >= 0x10000u) {
+    writeHex(
+        static_cast<char16_t>(
+            0xD800u + (((codePoint - 0x10000u) >> 10) & 0x3FFu)),
+        out);
+    writeHex(
+        static_cast<char16_t>(0xDC00u + ((codePoint - 0x10000u) & 0x3FFu)),
+        out);
+    return;
+  }
+  writeHex(static_cast<char16_t>(codePoint), out);
+  return;
+}
+
+void escapeString(const char* const input, size_t length, std::string& result) {
+  char buf[1 << 13];
+  // One unicode character would be encoded by at most 12 chars.
+  const size_t LOOK_AHEAD = 13;
+  char* pos = buf;
+
+  *pos++ = '"';
+
+  auto* start = reinterpret_cast<const unsigned char*>(input);
+  auto* end = reinterpret_cast<const unsigned char*>(input + length);
+  while (start < end) {
+    if (pos - buf + LOOK_AHEAD >= sizeof(buf)) {
+      flush(buf, pos, result);
+    }
+
+    size_t count = validateAndGetNextUtf8Length(start, end);
```

Comment:
Check out `utf8proc_iterate` in `velox/external/utf8proc/utf8proc.h`, you probably don't need to implement the validate/length/iteration yourselves.

Reply:
Hi Jimmy, I had gone over the funcs in utf8proc.h once, but found no function meeting all of needs.

Reply:
`utf8proc_iterate` is almost a drop-in replacement for `validateAndGetNextUtf8Length`.  What is not meeting our need here?

Reply:
`utf8proc_iterate()` also includes computation of the codepoint value, which we do not need here and thus unnecessary overhead. As we've already implemented validateAndGetNextUtf8Length(), I think we don't need to look further for an alternative.